nuxeo-features/nuxeo-automation/nuxeo-automation-core/src/main/java/org/nuxeo/ecm/automation/core/operations/blob/BlobToPDF.java

```diff
@@ -74,38 +86,96 @@ public Blob run(DocumentModel doc) {
+                String filename = blob.getFilename();
+                blob = convertBlobToMimeType(blob, MediaType.TEXT_HTML);
+                blob.setFilename(filename);
+            }
+            // Replace the image URLs by absolute paths
+            Path tempDirectory = Framework.createTempDirectory("blobs");
+            blob = replaceURLsByAbsolutePaths(blob, tempDirectory);
+            // Convert the blob to PDF
+            result = convertBlobToMimeType(blob, MimetypeRegistry.PDF_MIMETYPE);
+            org.apache.commons.io.FileUtils.deleteQuietly(tempDirectory.toFile());
+        }
         adjustBlobName(blob, result);
         return result;
     }
 
     @OperationMethod
-    public BlobList run(BlobList blobs) {
+    public BlobList run(BlobList blobs) throws IOException {
         BlobList bl = new BlobList();
         for (Blob blob : blobs) {
             bl.add(this.run(blob));
         }
         return bl;
     }
 
+    protected Blob convertBlobToMimeType(Blob blob, String mimetype) {
+        BlobHolder bh = new SimpleBlobHolder(blob);
+        bh = service.convertToMimeType(mimetype, bh, new HashMap<String, Serializable>());
+        Blob result = bh.getBlob();
+        return result;
+    }
+
     protected void adjustBlobName(Blob in, Blob out) {
         String fname = in.getFilename();
         if (fname == null) {
             fname = "Unknown_" + System.identityHashCode(in);
         }
         out.setFilename(fname + ".pdf");
-        out.setMimeType("application/pdf");
+        out.setMimeType(MimetypeRegistry.PDF_MIMETYPE);
+    }
+
+    /**
+     * Replace the image URLs of an HTML blob by absolute local paths
+     *
+     * @throws IOException
+     * @since 9.1
+     */
+    protected Blob replaceURLsByAbsolutePaths(Blob blob, Path tempDirectory) throws IOException {
+        String initialBlobContent = blob.getString();
+        String blobContentWithAbsolutePaths = initialBlobContent;
+        DownloadService downloadService = Framework.getService(DownloadService.class);
+        // Find images links in the blob
+        Pattern pattern = Pattern.compile("src=\"(.*?)\"");
+        Matcher matcher = pattern.matcher(initialBlobContent);
+        while (matcher.find()) {
+            // Retrieve the image from the URL
+            String url = matcher.group(1);
+            Blob imageBlob = downloadService.resolveBlobFromDownloadUrl(url);
+            if (imageBlob == null) {
+                break;
+            }
+            // Export the image to a temporary directory in File System
+            String safeFilename = FileUtils.getSafeFilename(imageBlob.getFilename());
+            File imageFile = tempDirectory.resolve(safeFilename).toFile();
+            imageBlob.transferTo(imageFile);
+            // Replace the image's URL by its absolute local path
+            blobContentWithAbsolutePaths = blobContentWithAbsolutePaths.replaceAll(url, imageFile.toPath().toString());
```

Comment:
It would be better to use [Matcher.appendReplacement](https://docs.oracle.com/javase/8/docs/api/java/util/regex/Matcher.html#appendReplacement-java.lang.StringBuffer-java.lang.String-) to only scan the string once during the iterative matching. See the example use in that Javadoc. You can keep using `matcher.group(1)` to get the current match group.